packages/upload-api/src/test/external-service/router.js

```diff
@@ -30,11 +31,34 @@ export const create = (serviceID, storageProviders) =>
         provider = undefined
       }
       if (!provider) {
-        provider = storageProviders[getRandomInt(storageProviders.length)].id
+        provider = storageProviders[getRandomInt(storageProviders.length - 1)].id
         stickySelect.set(key, provider)
       }
       return ok(provider)
     },
+    selectReplicationProviders: async (primary, count, digest, size, options) => {
+      const exclusions = [primary, ...(options?.exclude ?? [])]
+        .map(p => p.did())
+      const filteredProviders = storageProviders
+        .map(sp => sp.id)
+        .filter(id => !exclusions.includes(id.did()))
+
+      if (filteredProviders.length < count) {
+        return error(/** @type {API.BlobAPI.CandidateUnavailable} */ ({
+          name: 'CandidateUnavailable',
+          message: `Wanted ${count} but only ${filteredProviders.length} are available`
+        }))
+      }
+
+      /** @type {API.Principal[]} */
+      const selectedProviders = []
+      for (let i = 0; i < count; i++) {
+        const index = getRandomInt(filteredProviders.length - 1)
+        selectedProviders.push(filteredProviders[index])
+        filteredProviders.splice(index, 1)
+      }
+      return ok(selectedProviders)
+    },
```

Comment:
Nit: A code comment for this functions selection policy would be welcome!

Right now, it looks like we randomly select `count` unique storage providers to replicate data to and exclude the primary provider. In future versions of this I expect we'll want some round-robin type of selection policy?

Reply:
Yes, that's what will happen when we implement in w3infra, also taking into account weights. That logic will be tested in w3infra.

This implementation is for the tests here, where the selection logic is not known by the handlers and does not matter.